### ISSUE TITLE
update() not skipping null properties.

### ISSUE THREAD

**Issue by alessandrojcm** [typeorm/typeorm#931]
Hello, I have this entity defined:

```typescript
@Entity('users')
export class UserEntity {

  @PrimaryColumn('nvarchar', {unique: true, nullable: false})
  username: string

  @Column('nvarchar', {nullable: false})  
  name: string

  @Index('email', {unique: true})
  @Column('nvarchar', {unique: true, nullable: false})  
  email: string

  @Column('nvarchar', {nullable: false})  
  password: string

  @Column('nvarchar', {nullable: false})  
  role: UserRole
}
```

And then I have this method (simplified for clarity):
```typescript
async modifyUser (user: UserEntity): Promise<UserEntity> {
    if (user.email) {// This does not work.
          return this.connection.getRepository(UserEntity)
            .update({email: user.email}, user)
            .then(() => {
              return this.connection.getRepository(UserEntity).findOne({email: user.email})
            })
            .catch((error) => {
              this.logger.error(error)
              return Promise.reject(error)
            })
        } else if (user.username) {// This works.
          return this.connection.getRepository(UserEntity)
            .update({username: user.username}, user)
            .then(() => {
              return this.connection.getRepository(UserEntity).findOne({username: user.username})
            })
            .catch((error) => {
              this.logger.error(error)
              return Promise.reject(error)
            })
        } else {
          return Promise.reject('Es necesario buscar por email o username.')
        }
  }
```

It allows to search for an user using either it's username (the primary key) or the email (which is unique).
If I pass in the username it works fine, but if I pass the email I get this error:

```
RequestCannot insert the value NULL into column 'role', table 'testdb.dbo.users'; column does not allow nulls. INSERT fails.
```

Isn't `update()` supposed to do partial updates? Why it's the query an `INSERT` instead of an `UPDATE`?

**Comment by pleerock** [comment]
Yes. It actually (right now) calls `save` method which performs same - makes partial updates. Strange that you have `INSERT` query. I can only assume that you don't have a user with `{ email: user.email }` in the database. Please log what you send to `modifyUser` and check if you really have user with same email in the database. Also send me result of logs

**Comment by alessandrojcm** [comment]
I've just created an user and checked the database with DataGrip to ensure that it exists. Then I called the method.

This is the object as received by the method (sending the email and a new password):
```
{ email: 'testuser@testuser.com', password: '1234' }
Returning CLASS instance:
UserEntity {
  username: undefined,
  name: undefined,
  email: 'testuser@testuser.com',
  password: '1234',
  role: undefined }
```

This are the logs:
```
{
  "message": "Cannot insert the value NULL into column 'role', table 'AquasoftDev.dbo.users'; column does not allow nulls. INSERT fails.",
  "stack": "RequestError: Cannot insert the value NULL into column 'role', table 'AquasoftDev.dbo.users'; column does not allow nulls. INSERT fails.\n    at handleError (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\mssql\\lib\\tedious.js:519:15)\n    at emitOne (events.js:96:13)\n    at Connection.emit (events.js:188:7)\n    at Parser.<anonymous> (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\tedious\\lib\\connection.js:361:16)\n    at emitOne (events.js:96:13)\n    at Parser.emit (events.js:188:7)\n
at Parser.<anonymous> (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\tedious\\lib\\token\\token-stream-parser.js:54:15)\n    at emitOne (events.js:96:13)\n    at Parser.emit (events.js:188:7)\n    at addChunk (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\tedious\\node_modules\\readable-stream\\lib\\_stream_readable.js:284:12)\n    at readableAddChunk (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\tedious\\node_modules\\readable-stream\\lib\\_stream_readable.js:271:11)",
  "code": "EREQUEST",
  "number": 515,
  "lineNumber": 1,
  "state": 2,
  "class": 16,
  "serverName": "DESKTOP-LE9AE42\\SQLEXPRESS",
  "procName": "",
  "originalError": {
    "info": {
      "number": 515,
      "state": 2,
      "class": 16,
      "message": "Cannot insert the value NULL into column 'role', table 'AquasoftDev.dbo.users'; column does not allow nulls. INSERT fails.",
      "serverName": "DESKTOP-LE9AE42\\SQLEXPRESS",
      "procName": "",
      "lineNumber": 1,
      "name": "ERROR",
      "event": "errorMessage"
    }
  },
  "name": "RequestError",
  "precedingErrors": [],
  "level": "error",
  "timestamp": "2017-09-23T05:28:25.237Z"
}
{
  "message": "RequestCannot insert the value NULL into column 'role', table 'AquasoftDev.dbo.users'; column does not allow nulls. INSERT fails.",
  "stack": "RequestError: Cannot insert the value NULL into column 'role', table 'AquasoftDev.dbo.users'; column does not allow nulls. INSERT fails.\n    at handleError (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\mssql\\lib\\tedious.js:519:15)\n    at emitOne (events.js:96:13)\n    at Connection.emit (events.js:188:7)\n    at Parser.<anonymous> (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\tedious\\lib\\connection.js:361:16)\n    at emitOne (events.js:96:13)\n    at Parser.emit (events.js:188:7)\n
at Parser.<anonymous> (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\tedious\\lib\\token\\token-stream-parser.js:54:15)\n    at emitOne (events.js:96:13)\n    at Parser.emit (events.js:188:7)\n    at addChunk (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\tedious\\node_modules\\readable-stream\\lib\\_stream_readable.js:284:12)\n    at readableAddChunk (C:\\repos\\aquasoft\\aquasoft\\backend-server\\node_modules\\tedious\\node_modules\\readable-stream\\lib\\_stream_readable.js:271:11)",
  "code": "EREQUEST",
  "number": 515,
  "lineNumber": 1,
  "state": 2,
  "class": 16,
  "serverName": "DESKTOP-LE9AE42\\SQLEXPRESS",
  "procName": "",
  "originalError": {
    "info": {
      "number": 515,
      "state": 2,
      "class": 16,
      "message": "Cannot insert the value NULL into column 'role', table 'AquasoftDev.dbo.users'; column does not allow nulls. INSERT fails.",
      "serverName": "DESKTOP-LE9AE42\\SQLEXPRESS",
      "procName": "",
      "lineNumber": 1,
      "name": "ERROR",
      "event": "errorMessage"
    }
  },
  "name": "QueryFailedError",
  "precedingErrors": [],
  "query": "INSERT INTO \"users\"(\"email\", \"password\")  VALUES (@0,@1)",
  "parameters": [
    {
      "value": "testuser@testuser.com",
      "type": "nvarchar",
      "params": []
    },
    {
      "value": "$2a$15$9TkyF9keLbwCxETNqaa8EuC0n36O4PZmQQW6DOVVGBYEA9H9UO0i.",
      "type": "nvarchar",
      "params": []
    }
  ],
  "level": "error",
  "timestamp": "2017-09-23T05:28:25.246Z"
}
info:  RequestError: Cannot insert the value NULL into column 'role', table 'AquasoftDev.dbo.users'; column does not allow nulls. INSERT fails.
    at handleError (C:\repos\aquasoft\aquasoft\backend-server\node_modules\mssql\lib\tedious.js:519:15)
    at emitOne (events.js:96:13)
    at Connection.emit (events.js:188:7)
    at Parser.<anonymous> (C:\repos\aquasoft\aquasoft\backend-server\node_modules\tedious\lib\connection.js:361:16)
    at emitOne (events.js:96:13)
    at Parser.emit (events.js:188:7)
    at Parser.<anonymous> (C:\repos\aquasoft\aquasoft\backend-server\node_modules\tedious\lib\token\token-stream-parser.js:54:15)
    at emitOne (events.js:96:13)
    at Parser.emit (events.js:188:7)
    at addChunk (C:\repos\aquasoft\aquasoft\backend-server\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:284:12)
    at readableAddChunk (C:\repos\aquasoft\aquasoft\backend-server\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:271:11)
RequestError: Cannot insert the value NULL into column 'role', table 'AquasoftDev.dbo.users'; column does not allow nulls. INSERT fails.
    at handleError (C:\repos\aquasoft\aquasoft\backend-server\node_modules\mssql\lib\tedious.js:519:15)
    at emitOne (events.js:96:13)
    at Connection.emit (events.js:188:7)
    at Parser.<anonymous> (C:\repos\aquasoft\aquasoft\backend-server\node_modules\tedious\lib\connection.js:361:16)
    at emitOne (events.js:96:13)
    at Parser.emit (events.js:188:7)
    at Parser.<anonymous> (C:\repos\aquasoft\aquasoft\backend-server\node_modules\tedious\lib\token\token-stream-parser.js:54:15)
    at emitOne (events.js:96:13)
    at Parser.emit (events.js:188:7)
    at addChunk (C:\repos\aquasoft\aquasoft\backend-server\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:284:12)
    at readableAddChunk (C:\repos\aquasoft\aquasoft\backend-server\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:271:11)
```

**Comment by pleerock** [comment]
Okay, I see the issue. Since `update` calls same `save` method it only updates entity based on its primary key property. And since you are changing a primary key it thinks are you trying to insert a new entity. Its a bug, I'll need to change how `update` method works in repository. Until that try to use UpdateQueryBuilder:


```ts
        await connection
            .createQueryBuilder()
            .update(User)
            .set({ email: "new@email", ... })
            .where("email = :email", { email: "old@email" })
            .execute();
```

**Comment by alessandrojcm** [comment]
Excellent, thanks a lot for the workaround. Keep up the great work.

**Comment by pleerock** [comment]
`update` method was fixed and now it uses update query builder inside. Fix will be released in `0.1.1`
